addons/survey/controllers/main.py

```diff
@@ -263,179 +254,100 @@ def survey_display_page(self, survey_token, answer_token, prev=None, **post):
         if not answer_sudo.is_time_limit_reached:
+            # Prepare answers and comment by question
+            prepared_questions = {}
             for question in questions:
-                answer_tag = "%s_%s" % (survey_sudo.id, question.id)
-                errors.update(question.validate_question(post, answer_tag))
+                answer = post.get(str(question.id))
+                answer, comment = self._extract_comment_from_answers(question, answer)
+                prepared_questions[question.id] = {'answer': answer, 'comment': comment}
 
-        ret = {}
-        if len(errors):
-            # Return errors messages to webpage
-            ret['errors'] = errors
-        else:
-            if not answer_sudo.is_time_limit_reached:
-                for question in questions:
-                    answer_tag = "%s_%s" % (survey_sudo.id, question.id)
-                    request.env['survey.user_input_line'].sudo().save_lines(answer_sudo.id, question, post, answer_tag)
-
-            vals = {}
-            if answer_sudo.is_time_limit_reached or survey_sudo.questions_layout == 'one_page':
-                go_back = False
-                answer_sudo._mark_done()
-            elif 'button_submit' in post:
-                go_back = post['button_submit'] == 'previous'
-                next_page, last = request.env['survey.survey'].next_page_or_question(answer_sudo, page_or_question_id, go_back=go_back)
-                vals = {'last_displayed_page_id': page_or_question_id}
-
-                if next_page is None and not go_back:
-                    answer_sudo._mark_done()
-                else:
-                    vals.update({'state': 'skip'})
+            # Questions Validation
+            errors = {}
+            for question in questions:
+                answer = prepared_questions[question.id]['answer']
+                comment = prepared_questions[question.id]['comment']
+                errors.update(question.validate_question(answer, comment))
+            if errors:
+                return {'error': 'validation', 'fields': errors}
 
-            if 'breadcrumb_redirect' in post:
-                ret['redirect'] = post['breadcrumb_redirect']
+            # Submitting questions
+            for question in questions:
+                answer = prepared_questions[question.id]['answer']
+                comment = prepared_questions[question.id]['comment']
+                request.env['survey.user_input_line'].sudo().save_lines(answer_sudo.id, question, answer, comment)
+
+        if answer_sudo.is_time_limit_reached or survey_sudo.questions_layout == 'one_page':
+            answer_sudo._mark_done()
+        elif 'breadcrumb_page_id' in post:
+            # Go back to specific page using the breadcrumb. Lines are saved and survey continues
+            return '/survey/fill/%s/%s?breadcrumb_page_id=%s' % (survey_sudo.access_token, answer_token, post['breadcrumb_page_id'])
+        elif 'button_submit' in post:
```

Comment:
Could this be a simple 'else' ?
It should be the default case IMO.

And you would not even need to add 'button_submit' in the JS.

Reply:
Goed gezien dude ! (hey dude)